src/conversation_engine/engine.py

```diff
@@ -0,0 +1,409 @@
+
+        self._state = ConversationState()
+
+        self._action_client = Client(robot_name)
+
+        self._parser = cfgparser.CFGParser.fromstring(self._knowledge.grammar)
+        self._robot_name = robot_name
+
+        self._user_to_robot_sub = rospy.Subscriber("user_to_robot", String, self._handle_user_to_robot)
+        self._robot_to_user_pub = rospy.Publisher("robot_to_user", String, queue_size=10)
+
+        self._latest_feedback = None
+
+        rospy.logdebug("Started conversation engine")
+
+    def __del__(self):
+        self._stop()
+
+    def _handle_user_to_robot(self, msg):
+        rospy.loginfo("_handle_user_to_robot('{}')".format(msg))
+
+        text = sanitize_text(msg.data)
+
+        if self._handle_special_commands(text):
+            return
+
+        if self._state.state == ConversationState.IDLE:
+            # parse command and send_goal
+            self._handle_command(text)
+        elif self._state.state == ConversationState.WAIT_FOR_USER:
+            # Update semantics and send updated goal
+            self._handle_additional_info(text)
+        elif self._state.state == ConversationState.WAIT_FOR_ROBOT:
+            # User must wait for robot/action server to reply, cannot handle user input now
+            self._handle_user_while_waiting_for_robot(text)
+        elif self._state.state == ConversationState.ABORTING:
+            self._handle_user_while_aborting(text)
+
+    def _handle_special_commands(self, text):
+        """Check for special commands that should not be parsed further
+        @returns bool indicating if the text is a special command"""
+
+        stop_words = ["stop", "cancel", "quit", "reset"]
+        if any([word for word in stop_words if word in text]):
+            rospy.loginfo("_handle_special_commands('{}'):".format(text))
+            if self._state.state == ConversationState.IDLE:
+                self._robot_to_user_pub.publish(random.choice(["I'm not busy",
+                                                               "I'm not doing anything"]))
+                self._say_ready_for_command()
+            elif self._state.state == ConversationState.WAIT_FOR_ROBOT:
+                self._stop()
+            elif self._state.state == ConversationState.WAIT_FOR_USER:
+                self._robot_to_user_pub.publish(random.choice(["I'm waiting for you, there's nothing to stop",
+                                                               "I can't stop, you stop!"]))
+
+                self._start_new_conversation()
+            elif self._state.state == ConversationState.ABORTING:
+                self._robot_to_user_pub.publish(random.choice(["I'm already stopping, gimme some time!"]))
+            return True
+
+        kill_words = ["sudo kill"]
+        if any([word for word in kill_words if word in text]):
+            rospy.loginfo("_handle_special_commands('{}'): kill action server".format(text))
+
+            self._action_client.cancel_all_async()
```

Comment:
Does this call gives the action server sufficient time to handle the cancel request? If you kill it immediately, cancel requests might not have been passed on to the underlying action servers. E.g., if you kill the action server before it has processed the preempt request and sent the cancel request to the local planner server...

Reply:
This is a last resort, after you've already tried to send a 'stop' that should have cancelled all goals already. If you need the 'sudo kill', you're out of luck already. So maybe not even do the cancel_all here?